pkg/backends/tcp.go

```diff
@@ -152,10 +155,12 @@ func (ns *TCPSession) Recv() ([]byte, error) {
 
 // Close closes the session
 func (ns *TCPSession) Close() error {
-	if ns.closeChan != nil {
-		close(ns.closeChan)
-		ns.closeChan = nil
-	}
+	ns.once.Do(func() {
+		if ns.closeChan != nil {
```

Comment:
I feel like this logic should be inverted - we should check for nil-ness first, then wrap only the actual `close` call in the `sync.Once`.